sdk/monitor/Azure.Monitor.Ingestion/src/LogsIngestionClient.cs

```diff
@@ -106,51 +111,69 @@ internal static IEnumerable<BatchedLogs<T>> Batch<T>(IEnumerable<T> logEntries,
+                    writer.WriteEndArray();
+                    writer.Flush();
+                    // This batch is full so send it now
+                    yield return new BatchedLogs<T>(currentLogList, BinaryData.FromBytes(arrayBuffer.WrittenMemory));
+
+                    // Reset arrayBuffer and writer for next batch
+                    arrayBuffer = new Azure.Core.ArrayBufferWriter<byte>(SingleUploadThreshold);
+                    writer.Reset(arrayBuffer);
+                    writer.WriteStartArray();
+                    // reset log list
+                    currentLogList = new List<T>();
+                    // add current log to memory and currentLogList
+                    WriteMemory(writer, memory);
                     currentLogList.Add(log);
+
+                    // if this is the last log, send batch now
+                    if (isLastEntry)
+                    {
+                        writer.WriteEndArray();
+                        writer.Flush();
+                        yield return new BatchedLogs<T>(currentLogList, BinaryData.FromBytes(arrayBuffer.WrittenMemory));
+                    }
                 }
                 else
                 {
-                    WriteMemory(stream, memory);
-                    if ((entryCount + 1) == logEntries.Count())
-                    {
-                        // reached end of logs and we haven't returned yet
-                        WriteMemory(stream, BinaryData.FromString("]").ToMemory());
-                        stream.Position = 0;
-                        currentLogList.Add(log);
-                        yield return new BatchedLogs<T>(currentLogList, BinaryData.FromStream(stream));
-                    }
-                    else
+                    // Add entry to existing stream and update logList
+                    WriteMemory(writer, memory);
+                    currentLogList.Add(log);
+
+                    // if this is the last log, send batch now
+                    if (isLastEntry)
                     {
-                        WriteMemory(stream, BinaryData.FromString(",").ToMemory());
-                        currentLogList.Add(log);
+                        writer.WriteEndArray();
+                        writer.Flush();
+                        yield return new BatchedLogs<T>(currentLogList, BinaryData.FromBytes(arrayBuffer.WrittenMemory));
                     }
                 }
                 entryCount++;
             }
         }
 
-        private static void WriteMemory(MemoryStream stream, ReadOnlyMemory<byte> memory)
+        private static void WriteMemory(Utf8JsonWriter writer, ReadOnlyMemory<byte> memory)
         {
-            stream.Write(memory.ToArray(), 0, memory.Length); //TODO: fix ToArray
+            using (JsonDocument doc = JsonDocument.Parse(memory))
+            {
+                // Comma separator added automatically by JsonDocument
```

Comment:
nit: the separator is added by Utf8JsonWriter. Would be fine to just remove this comment as well.